### PR TITLE
Add a command to invalidate the LDAP cache

### PR DESCRIPTION
**Pull request by jvillafanez**
Link to https://github.com/owncloud/enterprise/issues/4533

The command will remove ALL the cache entries associated to the LDAP app.

### CODE REVIEW

**Comment by mmattel** [comment]
Docs relevant, pls file an issue when close to merge.
@EParzefall fyi

**Comment by jvillafanez** [comment]
There are no tests for any of the commands nor the `User_Proxy` nor `Group_Proxy`

File: lib/Command/InvalidateCache.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
+<?php
+/**
+ * @author Arthur Schiwon <blizzz@arthur-schiwon.de>
+ * @author Joas Schilling <coding@schilljs.com>
+ * @author Morris Jobke <hey@morrisjobke.de>
+ *
+ * @copyright Copyright (c) 2016, ownCloud GmbH.
+ * @license AGPL-3.0
```
Comment (JanAckermann) [inline lib/Command/InvalidateCache.php:8]: You should be here

**Review by JanAckermann — APPROVED** [review]
Code LGTM, thinking about some more tests, at least a command unit test

**Comment by JanAckermann** [comment]
Restarting tests

File: lib/Command/InvalidateCache.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	private $uProxy;
+	/** @var Group_Proxy */
+	private $gProxy;
+
+	/** @var UserMapping */
+	protected $mapping;
+	/**
+	 * @var ICache
+	 */
+	private $cache;
```
Comment (JammingBen) [inline lib/Command/InvalidateCache.php:46]: These vars can be removed because unused I believe. Also some imports above are unused.

**Review by JammingBen — CHANGES_REQUESTED** [review]
Tiny thing, rest LGTM

**Review by JammingBen — APPROVED** [review]

**Comment by jvillafanez** [comment]
Doc ticket opened in https://github.com/owncloud/docs/issues/3901

**Comment by voroyam** [comment]
```
root@oc:/var/www/owncloud# occ ldap:invalidate-cache carlos


  Too many arguments, expected arguments "command".


ldap:invalidate-cache

root@oc:/var/www/owncloud# occ ldap:invalidate-cache --help
Description:
  invalidates the cache for given users or all users.

Usage:
  ldap:invalidate-cache

Options:
  -h, --help            Display this help message
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --ansi            Force ANSI output
      --no-ansi         Disable ANSI output
  -n, --no-interaction  Do not ask any interactive question
      --no-warnings     Skip global warnings, show command output only
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug
root@oc:/var/www/owncloud# occ ldap:invalidate-cache
 4/4 [============================] 100%root@oc:/var/www/owncloud#
```

**Comment by voroyam** [comment]
Command works:

```
root@oc:/var/www/owncloud# occ user:list-groups carlos
  - g2
  - g1
root@oc:/var/www/owncloud# occ group:list-members g1
  - carlos: carlos
  - oc1: oc1 oc1
 ```

```
root@oc:/var/www/owncloud# occ ldap:invalidate-cache
 4/4 [============================] 100%
```

```
root@oc:/var/www/owncloud# occ group:list-members g1
  - oc1: oc1 oc1
  - oc2: oc2 oc2
root@oc:/var/www/owncloud# occ user:list-groups carlos
  - g2
 ```

**Comment by erwinpalma** [comment]
I have one question about the patch that solves this issue,

This patch will be present in the oc 10.8? os the customer need to patch it after the oc upgrade?

**Comment by mmattel** [comment]
If an updated LDAP version will be published on the marketplace, we need to add that command also at least to 10.8. A docs issue has been created: https://github.com/owncloud/docs/issues/3901. Please add a note in the dos issue  regarding the plans.

**Comment by jvillafanez** [comment]
It isn't bound to any particular OC version, so you need whatever minimum version the user_ldap app currently requires (I think it's 10.4)

**Comment by erwinpalma** [comment]
OK, @jvillafanez,

So, if the customer has oc10.8 then does he need to apply the patch anyway?

**Comment by jvillafanez** [comment]
Once the new version of the user_ldap app is released, he can install that version.
A patch for the app might work (not tested), but I'd recommend to wait / push for the new app version.

**Comment by enbrnz** [comment]
Is there also an API endpoint? Then you could configure a trigger on the LDAP backend for changes in the groups available in ownCloud to call that API endpoint.

**Comment by jvillafanez** [comment]
> Is there also an API endpoint?

Not planned as far as I'm aware of.
I don't know if it's easy or not to setup that feature in the LDAP backend, or how many providers have that kind of support.

**Comment by enbrnz** [comment]
In Windows AD all you have to do is enable the audit log, then you can just run `attach task to this event.`

**Pull request merged**